node/core/dispute-coordinator/src/real/ordering.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{
+	cmp::{Ord, Ordering, PartialOrd},
+	collections::{BTreeMap, HashMap, HashSet},
+};
+
+use polkadot_node_subsystem::{ActivatedLeaf, ActiveLeavesUpdate, SubsystemSender};
+use polkadot_node_subsystem_util::runtime::get_candidate_events;
+use polkadot_primitives::v1::{BlockNumber, CandidateEvent, CandidateHash, Hash, Id};
+
+use super::error::Result;
+
+/// Provider of `CandidateComparator` for candidates.
+pub struct OrderingProvider {
+	/// Currently cached comparators for candidates.
+	cached_comparators: HashMap<CandidateHash, CandidateComparator>,
+	/// including block -> `CandidateHash`
+	///
+	/// We need this to clean up `cached_comparators` on `ActiveLeavesUpdate`.
+	candidates_by_block_number: BTreeMap<BlockNumber, HashSet<CandidateHash>>,
+}
+
+/// Comparator for ordering of disputes for candidates.
+///
+/// This comparator makes it possible to order disputes based on age and to ensure some fairness
+/// between chains in case of equally old disputes.
+///
+/// Objective ordering between nodes is important in case of lots disputes, so nodes will pull in
+/// the same direction and work on resolving the same disputes first. This ensures that we will
+/// conclude some disputes, even if there are lots of them. While any objective ordering would
+/// suffice for this goal, ordering by age ensures we are not only resolving disputes, but also
+/// resolve the oldest one first, which are also the most urgent and important ones to resolve.
+///
+/// Note: That by `oldest` we mean oldest in terms of relay chain block number, for any block
+/// number that has not yet been finalized. If a block has been finalized already it should be
+/// treated as low priority when it comes to disputes, as even in the case of a negative outcome,
+/// we are already too late. The ordering mechanism here serves to prevent this from happening in
+/// the first place.
+#[derive(Copy, Clone)]
+pub struct CandidateComparator {
```

Comment:
In the other two comments I suggest using `relay_parent_number` and `relay_parent_hash` instead of `included_block_X`. But `relay_parent ++ para_id` is also not unique since multiple collators can build a parachain block. I suggest `(relay_parent_number, candidate_hash)` as the comparator.